### DISCUSSION TITLE
Game ideas

### DISCUSSION THREAD

**Discussion by AbdelStark** [keep-starknet-strange/tsubasa#2] [Ideas]
# Onchain Football Simulation Game

## Game Overview

This game is a football simulation game built on blockchain technology. Players are represented as NFTs with unique metadata. The game leverages smart contracts to manage gameplay and player interactions.

### Frontend

The frontend provides an interactive interface for players to manage their teams, view match statistics, and interact with other players. It displays the NFTs and their metadata in a user-friendly format.

### Onchain Smart Contracts

The smart contracts manage the game logic, player transactions, and interactions between NFTs. They ensure the game is fair, transparent, and secure.

### NFT Players

Each player is represented as an NFT with unique metadata, including attributes like speed, strength, skill, and special powers. These attributes affect the player's performance in matches.

### Team Formation

Players can form their own football teams by acquiring NFT players. They can trade, buy, or sell these NFT players on the marketplace.

### Matches

Matches are simulated based on the attributes of the NFT players in each team. The outcome of each match is determined by a combination of player attributes, strategy, and a bit of luck.

### Special Powers

Each NFT player has a unique special power that can be used once per match. These powers can turn the tide of a match, adding an extra layer of strategy to the game.

## Game Rules

1. Each team must have a minimum and maximum number of players as defined by the game rules.
2. Matches are simulated at specific times. Players must set their team lineup and strategy before the match begins.
3. The outcome of each match is determined by the smart contract, which takes into account the attributes of each NFT player and the chosen strategy.
4. Each NFT player's special power can be used once per match. The timing and use of these powers can greatly affect the match outcome.

## Expanded Game Rules

1. **Team Formation**: Each team must consist of 11 players, including one goalkeeper, at least three defenders, at least three midfielders, and at least one forward. The remaining players can be of any position. Players can adjust their team formation before each match.

2. **Match Setup**: Matches are simulated at specific times, announced in advance. Players must set their team lineup and strategy before the match begins. The strategy can include the aggressiveness of play (defensive, balanced, or offensive), the focus of attacks (wings or center), and the timing of special powers.

3. **Match Simulation**: The outcome of each match is determined by the smart contract, which takes into account the attributes of each NFT player, the chosen strategy, and a random factor to simulate the unpredictability of football.

4. **Special Powers**: Each NFT player's special power can be used once per match. The timing and use of these powers can greatly affect the match outcome. Special powers can include "Speed Boost", "Precision Shot", "Iron Defense", etc.

5. **Player Transactions**: Players can trade, sell, or buy NFT players on the marketplace. All transactions are managed by smart contracts to ensure fairness and security.

## Game Example

1. **Team Formation**: Alice forms her team with a balanced mix of defenders, midfielders, and forwards. She has a star player with a "Precision Shot" special power.

2. **Match Setup**: A match is announced to take place in 24 hours. Alice sets her strategy to "balanced" and decides to focus her attacks on the wings. She plans to use her star player's special power in the second half of the match. The match can also be triggered manually or at a specific block.

3. **Match Simulation**: The match begins. The smart contract simulates the match based on Alice's strategy, her players' attributes, and a random factor. At half-time, the score is 0-0.

4. **Special Powers**: In the second half, Alice activates her star player's "Precision Shot" special power. The smart contract takes this into account in the match simulation. A few minutes later, Alice's team scores a goal.

5. **End of Match**: The match ends with a 1-0 victory for Alice's team. Alice receives rewards based on her victory, which she can use to improve her team or trade on the marketplace.

**Comment by tarrencev** [comment]
High level this looks good. We should remove the emphasis on NFT / trading / etc for the demo. It can be implemented in simple dojo ecs.

> Matches are simulated at specific times

We should support async matches. Players create and submit a team to the tournament. Then you can play other teams in the tournament by simulating games against them on demand.

For MVP, do we want just two predefined teams?

For the simulation, I think we can use Orion to define a probabilistic model. It should take as input the attributes of the players on each side and predict a score as output. The model weights can be perturbed at runtime to add some randomness to the game

What attributes are defined from sorare players today?

**Comment by 0xLucqs** [comment]
**Proposition**

Let's do a simplified hearthstone like game

# General rules

* A game ends when a player wins 2 rounds.
* The first player to score a goal wins the round
* Each player has a deck of 11 cards (prepared before)
* At each round you start with 3 random cards from your deck
* At every turn of a round you get a new random card from your deck
* Each card requires a specified amount of energy to be placed on the board
* Each card has defense/attack stats
* At the end of a turn the player's card will attack the opponent's cards. If your card has `A` attack points and `B` defense points and the opponent's card has `X` attack points and `Y` defense points the opponent's card will receive `A` damage, if `Y - A <= 0` the enemy card will go out of the board. When attacking a card yours will also receive `X` damage. If `B - X <= 0` your card also goes out of the board.
*  If the opponent doesn't have any card on the board your card will shoot at the enemy goal. The probability that he scores the goal is equal to his attack statistics.
* You can only put your card on the leftmost/rightmost side of the board

# Note

We can add a mechanism where each player can bet 1 of his cards and the winner gets it.
In the future if we have enough time we can add powers to each cards and think of other mechanism but this is out of the scope for the MVP.

**Comment by notV4l** [comment]
# General game ideas, and bla bla

mostly some mid/long term ideas, but idea is to restrict it do define to an initial scope keeping in mind what could come next

## Join the game

You join the game by minting a 'Starter Deck'
Its limited to 1 by wallet and contains :

- Your **Captain** (team captain, its a player, its your profile)
- 5/10/15/?? extra players cards from different *teams*

- join the game on friend invite and get 3/5/?? extra cards

## Captain Card

The **Captain** is your **unique** in-game avatar, your representation in the game.
It should reflect what you like ( art / name), make you feel evolving (customizable art / xp / lvl / rewards)
It accrue all your game stats, XP, LVL, high facts, cups, etc

It's also your team Captain of course and take part of each game you play.

- art should be customizable / update with xp/lvl
- name (can be defined once for free, then can change with *xp token*)
- 1 initial *team* (can change with *xp token*, lvl)
- 1 initial *role* (can change with *xp token*, lvl)
- *attributes*
- 1 initial *special power* (can change with xp token, lvl)

note : 
- SBT ?
- *team* / *role* / *special power* : should be random at mint & customizable
- *attributes* : should be equal for all & be able to evolve


## Player Card

A card represent a football player from a *team*, a JPEG with cool attributes

- name (cant change)
- *team* (cant change)
- *role* (cant change)
- *attributes*
- *rarity* (cant change)
- *health*
- *state*

- playable or collector


## Playable / Collector / Collector+SBT 

Some peoples like to play, some like to collect, some other to trade... it can satisfy all at once

As the game evolve some mechanics are require to limit playable card numbers / balance the game
A card can be playable, then as cards/teams nb grow we should be able so turn some cards from *playable* to *collector*.
Some card could only be *collector* (more like a classic nft, but linked to the game, with same attributes than playable cards)

Another thing mainly for collectooors could be unique collector cards, special nft trophies/cups, ... 

## Teams

Like in football, teams are a set of (15/20/??) *player cards* playing under the same colors/art/values/etc
As a game player you'll always have your favorite team that you want to represent.
It should be wide enough so players can identify to what they likes

Team ideas :
- starknet ecosystem : Starkware, Dojo, Madara, ...
- crypto team : team BTC, team ETH, team STRK, ...
- ntf base : briq, everAI, ...
- event base : olympic games, halving, ...
- manga fan : dbz, naruto etc (probably with legal issues coming with XD )
- real football team with weird names like in old PESs XD, (Kylian Mbappé --> Pylian Mbakké)
- fantasy teams around game lore / ambiance
- infinite possibilities

Utility :
- playing with player from the same *team* can give team boost/special team effect

- community/involvment/fans
- trading/collection, (like panini albums, collect all cards to get the full team)
- collecting full team could make you earn a "team nft" (ie team banner)

Initial : 
- 4/5 teams around a common theme

Future : 
- release new teams regulary to keep the game evolving 
- new teams introduced could be determined by gamers gouvernance votes (= involvement)

## Card Roles & Attributes & Rarity

### Roles

There is some rules about *roles* to compose a team for a match, for ex at least one Goalkeeper.

- Goalkeeper
- Defender
- Midfielder
- Attacker

could be funnier with some custom fantasy names or fantasy roles

### Attributes

Different attributes :
- Attack
- Defense
- Speed
- Endurance
- ...

can be extend with other attributes

Each attribute has a value between ~~0-felt252~~ 0-100.

There is a :
- base value ( same for all similar cards ex: 69 for card C1 for team T1)
- current value : the current attribute value (between base value & max value)
- max value : defined by rarity

Trading a card reset current value to base value ?

The **captain** starts with all attributes values set to 85 and can evolve with xp/lvl up (skill points)

### Rarity / Card supply

Card rarity should be tied with card supply.
Rarity is visually recognizable when looking at cards, ad it affects *attributes* max values

- Unique           unique cards, much collectible, very wow

- Legendary +9    /  1% of supply
- Epic      +5    /  2% of supply
- Rare      +3    /  7% of supply
- Uncommon  +2    / 20% of supply
- Common    +1    / 70% of supply

ex for same card with diff rarity :
- Common    : base_value : 69, max_value : 70
- Rare      : base_value : 69, max_value : 72
- Legendary : base_value : 69, max_value : 78

Rarity should be reflect on card design to be easily recognizable, usually with some card border, visual effect, animations

To reach max value for attribute you have to earn xp, lvl, and use utility token.

- rewarding for players
- introduce rarity / small diff between 'same' card
- diff valuations

## Health & State/Status

### Health

Health is your player health, it a value 0-100.

It plays a role in players management, you cannot spam one player/one team and force you to make changes/choices in team composition.
Each game consume some health (ex: win: -10, even: -15, loss: -20)
Each player earn XX health /day (ex: 10/day or from 5 to 10 related to endurance, captain special power, etc)

90-100 : your player is at his top ( bonus for all attributes )
70-90 : normal stats (u should aim to only play with healthy players)
50-70 : performance drop (malus for all attributes, high % chances to get injured)
0-50 : u should not play but if u play u'll get injured u degen

### State / Status

A special state/status which affect your player card temporary.

Special events may happen when a player takes part to a game.
It could be game related event, or health related ( random or due to bad player managment )
It spice up the game and add a fun / strategy layer

Health related : 
- Injured -> you cannot play this card for 7 days
- Heart Attack -> rip / burn (extreme low health player)

Random:
- Sanctioned yellow/red -> you cannot play this card for 2/4 days

After victory:
- euphoria : your player feels so good it has some boosted *attributes* (+1/+2/+X on some for next match)
- fast recovery : the victory makes your player feel stronger and recover faster (recover x2 faster)

After defeat :
- Nervous breakdown -> you player really need to rest for 1 day
- Traumatised -> You got crushed so hard previous game this player cannot play for 3 days

Can evolve in multiple way

## XP, Level & Utility token

You should earn enough to feel rewarded, but you always want to get more XD, it must take time to unlock in-game content, high facts etc

### XP & Level

Any successful, addictive game use this common mechanics.
You earn XP mostly by playing games, there is xp steps where you level up as usual.
You always want to play one more game to level up, or get missing utility token u need to unlock/buy some in-game stuff
This is what keep you playing the game more and you feel rewarded by evolving, unlocking stuff

Other way to earn xp :
- accomplish in game high facts (win 5/10/15 games in a row, ...)
- missions ( trade at least 5 cards, play at least one game each day for XX days )
- Special events (extra XP, XP boost)
- win tournament, cups

Some minimum Level should be require to unlock some stuff for ex :

lvl  5 : unlock special powers
lvl 10 : unlock cards evolution
lvl 15 : unlock **captain** evolution
lvl 20 : unlock ability to gamble on games
.

Each level give X "skill point" to your **captain** that you can spend to evolve his attribute values  ( ex: +1 speed, +1 attack, ...)

### Utility token

This is in-game token, and can have lot of different utility.
Should not be tradable, tied to **Captain**
You earn it like XP, by playing game, doing missions, etc

- customize **Captain** ( change name, *role*, *team*, art)
- evolve card *attributes* values related with *rarity*
- mint cards boosters
- buy collector cards
- require an amount to lvl up
- [insert utility]


# THE GAME

...
